share/shwap/p2p/shrex/shrexnd/server.go

```diff
@@ -0,0 +1,219 @@
+	srv.host.SetStreamHandler(srv.protocolID, srv.handler)
+	return nil
+}
+
+// Stop stops the server
+func (srv *Server) Stop(context.Context) error {
+	srv.cancel()
+	srv.host.RemoveStreamHandler(srv.protocolID)
+	return nil
+}
+
+func (srv *Server) streamHandler(ctx context.Context) network.StreamHandler {
+	return func(s network.Stream) {
+		err := srv.handleNamespacedData(ctx, s)
+		if err != nil {
+			s.Reset() //nolint:errcheck
+			return
+		}
+		if err = s.Close(); err != nil {
+			log.Debugw("server: closing stream", "err", err)
+		}
+	}
+}
+
+// SetHandler sets server handler
+func (srv *Server) SetHandler(handler network.StreamHandler) {
+	srv.handler = handler
+}
+
+func (srv *Server) observeRateLimitedRequests() {
+	numRateLimited := srv.middleware.DrainCounter()
+	if numRateLimited > 0 {
+		srv.metrics.ObserveRequests(context.Background(), numRateLimited, shrex.StatusRateLimited)
+	}
+}
+
+func (srv *Server) handleNamespacedData(ctx context.Context, stream network.Stream) error {
+	logger := log.With("source", "server", "peer", stream.Conn().RemotePeer().String())
+	logger.Debug("handling nd request")
+
+	srv.observeRateLimitedRequests()
+	ndid, err := srv.readRequest(logger, stream)
+	if err != nil {
+		logger.Warnw("read request", "err", err)
+		srv.metrics.ObserveRequests(ctx, 1, shrex.StatusBadRequest)
+		return err
+	}
+
+	logger = logger.With(
+		"namespace", ndid.DataNamespace.String(),
+		"height", ndid.Height,
+	)
+
+	ctx, cancel := context.WithTimeout(ctx, srv.params.HandleRequestTimeout)
+	defer cancel()
+
+	nd, status, err := srv.getNamespaceData(ctx, ndid)
+	if err != nil {
+		// server should respond with status regardless if there was an error getting data
+		sendErr := srv.respondStatus(ctx, logger, stream, status)
+		if sendErr != nil {
+			logger.Errorw("sending response", "err", sendErr)
+			srv.metrics.ObserveRequests(ctx, 1, shrex.StatusSendRespErr)
+		}
+		logger.Errorw("handling request", "err", err)
+		return errors.Join(err, sendErr)
+	}
+
+	err = srv.respondStatus(ctx, logger, stream, status)
+	if err != nil {
+		logger.Errorw("sending response", "err", err)
+		srv.metrics.ObserveRequests(ctx, 1, shrex.StatusSendRespErr)
+		return err
+	}
+
+	_, err = nd.WriteTo(stream)
+	if err != nil {
+		logger.Errorw("send nd data", "err", err)
+		srv.metrics.ObserveRequests(ctx, 1, shrex.StatusSendRespErr)
+		return err
+	}
+	return nil
+}
+
+func (srv *Server) readRequest(
+	logger *zap.SugaredLogger,
+	stream network.Stream,
+) (shwap.NamespaceDataID, error) {
+	err := stream.SetReadDeadline(time.Now().Add(srv.params.ServerReadTimeout))
+	if err != nil {
+		logger.Debugw("setting read deadline", "err", err)
+	}
+
+	ndid := shwap.NamespaceDataID{}
+	_, err = ndid.ReadFrom(stream)
+	if err != nil {
+		return shwap.NamespaceDataID{}, fmt.Errorf("reading request: %w", err)
+	}
+
+	logger.Debugw("new request")
+	err = stream.CloseRead()
+	if err != nil {
+		logger.Debugw("closing read side of the stream", "err", err)
```

Comment:
If not error+reset, we should Warn here at least. This is abnormal and should be visible.

Reply:
The set deadline is a special case, because its coming from stupid mocknet where deadlines are not supported

Reply:
Same is true for other cases with CloseRead/Write